chips/litex_vexriscv/src/chip.rs

```diff
@@ -0,0 +1,215 @@
+            unsafe {
+                self.handle_interrupts();
+            }
+        }
+
+        // Re-enable all MIE interrupts that we care about. Since we
+        // looped until we handled them all, we can re-enable all of
+        // them.
+        CSR.mie.modify(mie::mext::SET);
+    }
+
+    fn has_pending_interrupts(&self) -> bool {
+        self.interrupt_controller.next_saved().is_some()
+    }
+
+    fn sleep(&self) {
+        unsafe {
+            rv32i::support::wfi();
+        }
+    }
+
+    unsafe fn atomic<F, R>(&self, f: F) -> R
+    where
+        F: FnOnce() -> R,
+    {
+        rv32i::support::atomic(f)
+    }
+
+    unsafe fn print_state(&self, writer: &mut dyn Write) {
+        let _ = writer.write_fmt(format_args!(
+            "\r\n---| LiteX configuration for {} |---",
+            self.soc_identifier,
+        ));
+        rv32i::print_riscv_state(writer);
+    }
+}
+
+fn handle_exception(exception: mcause::Exception) {
+    match exception {
+        mcause::Exception::UserEnvCall | mcause::Exception::SupervisorEnvCall => (),
+
+        mcause::Exception::InstructionMisaligned
+        | mcause::Exception::InstructionFault
+        | mcause::Exception::IllegalInstruction
+        | mcause::Exception::Breakpoint
+        | mcause::Exception::LoadMisaligned
+        | mcause::Exception::LoadFault
+        | mcause::Exception::StoreMisaligned
+        | mcause::Exception::StoreFault
+        | mcause::Exception::MachineEnvCall
+        | mcause::Exception::InstructionPageFault
+        | mcause::Exception::LoadPageFault
+        | mcause::Exception::StorePageFault
+        | mcause::Exception::Unknown => {
+            panic!("fatal exception");
+        }
+    }
+}
+
+unsafe fn handle_interrupt(intr: mcause::Interrupt) {
+    match intr {
+        mcause::Interrupt::UserSoft
+        | mcause::Interrupt::UserTimer
+        | mcause::Interrupt::UserExternal => {
+            debug!("unexpected user-mode interrupt");
+        }
+        mcause::Interrupt::SupervisorExternal
+        | mcause::Interrupt::SupervisorTimer
+        | mcause::Interrupt::SupervisorSoft => {
+            debug!("unexpected supervisor-mode interrupt");
+        }
+
+        mcause::Interrupt::MachineSoft => {
+            CSR.mie.modify(mie::msoft::CLEAR);
+        }
+        mcause::Interrupt::MachineTimer => {
+            CSR.mie.modify(mie::mtimer::CLEAR);
+        }
+        mcause::Interrupt::MachineExternal => {
+            // We received an interrupt, disable interrupts while we handle them
+            CSR.mie.modify(mie::mext::CLEAR);
+
+            // Save the interrupts and check whether at least one
+            // interrupt is to be handled
+            //
+            // If no interrupt was saved, reenable interrupts
+            // immediately
+            if !INTERRUPT_CONTROLLER.save_pending() {
+                CSR.mie.modify(mie::mext::SET);
+            }
+        }
```

Comment:
This should be done in a loop.

Reply:
I'm not quite sure I follow. On `e310x` for example, we run this in a loop as `save_interrupt(&self, index: u32)` on the PLIC saves only a single interrupt, whereas this `save_pending()` saves the entire interrupt register at once.

The only difference is that on the `e310x`, as soon as all interrupts are saved internally, the `mext` is enabled again, whereas on `litex_vexriscv` they're left disabled until all of them are handled (in line 96). What's the approach to use here?